File: foolish-core-java/src/main/java/org/foolish/fvm/ubc/SeekFiroe.java
package org.foolish.fvm.ubc;

import org.foolish.ast.AST;
import org.foolish.ast.SearchOperator;

/**
 * SeekFiroe implements offset-based access to a brane.
 * Syntax: b#0 (first), b#-1 (last).
 */
public class SeekFiroe extends AbstractSearchFiroe {

    private final int offset;

    public SeekFiroe(AST.SeekExpr seekExpr) {
        super(seekExpr, SearchOperator.SEEK);
        this.offset = seekExpr.offset();
    }

    /**
     * Copy constructor for cloneConstanic.
     */
    protected SeekFiroe(SeekFiroe original, FIR newParent) {
        super(original, newParent);
        this.offset = original.offset;
    }

    @Override
    protected void initialize() {
        super.initialize();
        AST.SeekExpr seekExpr = (AST.SeekExpr) ast;
        storeExprs(seekExpr.anchor());
    }

    @Override
    protected FIR executeSearch(Cursor cursor) {
        ReadOnlyBraneMemory targetMemory = cursor.brane().getBraneMemory();
        int size = targetMemory.size();
        int idx = offset;

        // Handle negative indexing
        if (idx < 0) {
            idx = size + idx;
        }

        // Check bounds
        if (idx >= 0 && idx < size) {
             return targetMemory.get(idx);
        }

        return new NKFiroe();
    }

    @Override
    public String toString() {
        return ast.toString();
    }

    @Override
    protected FIR cloneConstanic(FIR newParent, java.util.Optional<Nyes> targetNyes) {
        if (!isConstanic()) {
            throw new IllegalStateException(
                formatErrorMessage("cloneConstanic can only be called on CONSTANIC or CONSTANT FIRs, " +
                    "but this FIR is in state: " + getNyes()));
        }

        if (isConstant()) {
            return this;  // Share CONSTANT seeks
        }

        SeekFiroe copy = new SeekFiroe(this, newParent);

        if (targetNyes.isPresent()) {
            copy.nyes = targetNyes.get();
        } else {
            copy.nyes = this.nyes;
        }

        return copy;
    }
}

File: foolish-core-scala/src/main/scala/org/foolish/fvm/scubc/SeekFiroe.scala
package org.foolish.fvm.scubc

import org.foolish.ast.{AST, SearchOperator}

/**
 * SeekFiroe implements offset-based access to a brane.
 * Syntax: b#0 (first), b#-1 (last).
 */
class SeekFiroe(seekExpr: AST.SeekExpr) extends AbstractSearchFiroe(seekExpr, SearchOperator.SEEK) {

  private var offset: Int = seekExpr.offset()

  override protected def initialize(): Unit = {
    super.initialize()
    val expr = ast.asInstanceOf[AST.SeekExpr]
    enqueueExprs(expr.anchor())
  }

  override protected def executeSearch(target: BraneFiroe): FIR = {
    val targetMemory = target.braneMemory
    val size = targetMemory.size
    var idx = offset

    // Handle negative indexing
    if (idx < 0) {
      idx = size + idx
    }

    // Check bounds
    if (idx >= 0 && idx < size) {
      targetMemory.get(idx)
    } else {
      new NKFiroe()
    }
  }

  override def toString: String = ast.toString

  /**
   * Copy constructor for cloneConstanic.
   * Resets search state so the search can be re-executed in a new context.
   */
  private def this(original: SeekFiroe, newParent: FIR) =
    this(original.ast.asInstanceOf[AST.SeekExpr])
    setParentFir(newParent)
    offset = original.offset
    // Copy braneMemory contents with cloned items
    original.braneMemory.stream.foreach { fir =>
      val cloned = fir.asInstanceOf[Constanicable].cloneConstanic(this, Some(Nyes.INITIALIZED))
      braneMemory.put(cloned)
      indexLookup.put(cloned, braneMemory.size - 1)
      // For nested FiroeWithBraneMind instances, set up parent memory link
      if cloned.isInstanceOf[FiroeWithBraneMind] then
        cloned.asInstanceOf[FiroeWithBraneMind].ordinated = false
        cloned.asInstanceOf[FiroeWithBraneMind].ordinateToParentBraneMind(this, indexLookup.size - 1)
    }

  override def cloneConstanic(newParent: FIR, targetNyes: Option[Nyes]): FIR =
    if !isConstanic then
      throw IllegalStateException(
        s"cloneConstanic can only be called on CONSTANIC or CONSTANT FIRs, " +
        s"but this FIR is in state: ${getNyes}")
    if isConstant then
      return this  // Share CONSTANT seeks
    val copy = new SeekFiroe(this, newParent)
    targetNyes.foreach(copy.setNyes)
    copy
}
